### DISCUSSION TITLE
Is there a performance overhead when using MPh?

### DISCUSSION THREAD

**Discussion by shomikverma** [MPh-py/MPh#103] [General]
Hi,

I was comparing simulation speed of running a simple setup in MPh vs. in COMSOL, and found MPh was about 5x slower. Was curious if other people were facing this issue too.

I've uploaded my files [here](https://github.com/mary-foxen/graphite-modeling/tree/main): COMSOL file is graphite-thermal-conductivity-2D-noimport.mph and python is barebonestest.py.

COMSOL runs in ~1s while MPh takes around 7s. (This is just the solving function, not including client starting and loading etc.)

On a similar but separate note, I found if I ran the same simulation multiple times, subsequent runs were faster. Code at above repo, called simtimecomp.py.

![simtimes](https://user-images.githubusercontent.com/40326897/205421878-9de84c19-a67f-4179-a278-9d69ba1cf322.png)

Any insights?

I'm running on an M1 Mac with COMSOL 60a (Intel version)

**Comment by john-hen** [comment]
I cannot reproduce that. On my (much older) hardware, I get a "solution time" (as reported in the Messages tab of the Comsol GUI) of 8 to 9 seconds on the first run, and then 5 to 6 seconds on subsequent runs. With your script (`simtimescomp.py`), the average solution time is reported as about 4 seconds, so actually faster. Which makes sense: It doesn't have to update the GUI.

There is no reason to expect any meaningful differences in terms of computation times. After all, MPh is just a scripting interface, just like the Matlab LiveLink. The Comsol GUI is an interactive user interface. In the end, these interfaces all call the same (Java-based) computational back-end: the Comsol core, if you will.

There is a measurable difference between the first run and subsequent runs because, at first, Comsol also has to build the geometry and create the mesh. These two steps are not necessary when you solve the same model again, as they are cached. I don't think Comsol even clears that cache when you reload the model, it just reuses the same internal objects. A better benchmark would use a model with a solution time on the order of at least minutes, just so that the setup / teardown effects are negligible.
